### PR TITLE
Allow authenticating with a username to redis

### PR DESCRIPTION
**Pull request by maddymeows**
I needed this

I just looked for occurrences of password and added username branches there. Let me know if there's anything wrong, I'm not proficient in C at all.

### CODE REVIEW

**Comment by eakraly** [comment]
Thank you @maddymeows !
It seems it would allow coturn to work with Redis6 that has new ACL feature

Could you please rebase the PR and also explain if you tested with new and old redis.
Does it require minimal version of hiredis to support it?

**Comment by maddymeows** [comment]
Sorry for tagging the regression fix alongside this PR, came up in testing after rebasing.

> Could you please rebase the PR
Done

> and also explain if you tested with new and old redis.
I've tested on debian 10 which has redis dated before ACLs were introduced, works fine as long as you don't configure a username.
I've also tested on latest debian and fedora and works just as well

> Does it require minimal version of hiredis to support it?
As far as I know, no

File: src/apps/relay/dbdrivers/dbd_redis.c [hunk 1]
```diff
@@ -163,9 +167,6 @@ static Ryconninfo *RyconninfoParse(const char *userdb, char **errmsg) {
     if (!(co->host)) {
       co->host = strdup("127.0.0.1");
     }
-    if (!(co->password)) {
```
Comment (eakraly) [inline src/apps/relay/dbdrivers/dbd_redis.c:166]: Why this change?

Comment (maddymeows) [reply src/apps/relay/dbdrivers/dbd_redis.c:166]: Originally during testing I created an extra branch `if (!(co->user)) { co->user = strdup(""); }`.

However, I found out that it made coturn always authenticate with a username and password, which would break for redis servers that don't have the ACL feature.

That's because the branches using password checked if the pointer to password was set, which it always was with the statement I just removed, rather than if password contained a useful value. So the password conditionals haven't been working for a while, coturn just discarded the error that redis threw out and continued on (which is fine, but not ideal).

Hence also changes to make those branches `co->password && strlen(co->password)` rather than just `co->password`, which I could effectively change to `co->password && co->password[0]` but strlen felt nicer.

Error handling of missing auth doesn't seem to be handled fully correctly, but happens moreso as a side effect, the side effect of using the side effect is it never warns about if the password is needed which is why there are no logs complaining about use of `AUTH` when wasn't needed.

File: src/apps/relay/dbdrivers/dbd_redis.c [hunk 2]
```diff
@@ -268,7 +273,7 @@
         }
       }
 
-      ret = redisLibeventAttach(base, co->host, co->port, co->password, atoi(co->dbname));
+      ret = redisLibeventAttach(base, co->host, co->port, co->user, co->password, atoi(co->dbname));
```
[the remark this answers is not in the document]
Comment (maddymeows) [reply src/apps/relay/dbdrivers/dbd_redis.c:276]: `co->dbname` is never null because a default is given in `RyconninfoParse`, I don't know how to silence this.

**Comment by eakraly** [comment]
Hi @maddymeows ! Could you please update the PR so I can merge it into upcoming version? Thanks

**Comment by maddymeows** [comment]
@eakraly done!

**Comment by justinschoeman** [comment]
This commit breaks unauthenticated redis connections ("select <dbname>" is not sent if password is not set).

**Comment by maddymeows** [comment]
> This commit breaks unauthenticated redis connections ("select " is not sent if password is not set).

<https://github.com/coturn/coturn/pull/1668>
